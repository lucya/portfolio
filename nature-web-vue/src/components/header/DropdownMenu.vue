<template>
  <img :src="menuImg" alt='menu' ref="dropMenuRef" @click="handleOpen" />
  <div class='drop-menu' :style="menuOpen ? 'display: block' : 'display: none'">
    <ul>
      <li>
        <router-link to="/movies">인기영화</router-link>
      </li>
      <li>
        <router-link to="/fortune">운세마법사</router-link>
      </li>
      <li @click="handleLogout">로그아웃</li>
    </ul>
  </div>
</template>

<script>
import { ref, watch } from 'vue';
import http from '@/http-common'
import { useRouter } from 'vue-router'
import menuImg from '../../assets/images/menu_30.png';


export default {
  setup() {
    const router = useRouter()
    const dropMenuRef = ref(null)
    const menuOpen = ref(false)

    const handleOpen = () => {
      menuOpen.value = !menuOpen.value
    }

    const handleLogout = async () => {
      try {
        await http.post('/user/logout')
        router.replace({
          name: "Login",
        });
      } catch (error) {
        alert('서버 에러가 발생하고 있습니다.')
        console.log(error)
      }
    }
    watch([menuOpen.value], () => {
      const handleOutsideClose = (e) => {
        e.preventDefault();
        // useRef current에 담긴 엘리먼트 바깥을 클릭 시 드롭메뉴 닫힘
        if (menuOpen.value && (!dropMenuRef.value.current?.contains(e.target))) menuOpen.value = false
      }

      document.addEventListener('click', handleOutsideClose);
      return () => document.removeEventListener('click', handleOutsideClose);
    })

    return {
      dropMenuRef,
      handleLogout,
      handleOpen,
      menuOpen,
      menuImg,
    }
  }
}
</script>

<style></style>
