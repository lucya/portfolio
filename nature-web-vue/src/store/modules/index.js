import user from './user';
import movie from './movie';

export default {
  user,
  movie,
}