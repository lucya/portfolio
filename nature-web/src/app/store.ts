import { Action, configureStore, ThunkDispatch } from "@reduxjs/toolkit";
import logger from 'redux-logger'
import rootReducer from "../reducers";
import thunk from 'redux-thunk'
import {
  persistStore,
  PERSIST,
  PURGE
} from 'redux-persist'
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware(
      {
        serializableCheck: {
          ignoredActions: [PERSIST, PURGE],
        },
      }
    )
      .concat(logger).concat(thunk),
  devTools: process.env.MODE_ENV !== 'production',
});// redux store 정의

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>

export const persistor = persistStore(store); // redux store 생성

export default store;