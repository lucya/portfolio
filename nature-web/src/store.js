import { configureStore } from "@reduxjs/toolkit";
import logger from 'redux-logger';
import rootReducer from "./reducers";
import thunk from 'redux-thunk'
import { persistStore } from "redux-persist";


const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware()
    .concat(logger).concat(thunk),
  devTools: process.env.NODE_ENV !== 'production',
});// redux store 정의

export const persistor = persistStore(store); // redux store 생성

export default store;