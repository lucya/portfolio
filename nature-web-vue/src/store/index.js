import { createStore } from 'vuex'
import createPersistedState from 'vuex-persistedstate'; // 새로고침에서도 state 유지

import modules from './modules';

export default createStore({
  modules,
  plugins: [
    createPersistedState()
  ]
})
