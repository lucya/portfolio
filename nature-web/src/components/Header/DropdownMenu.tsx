import { useEffect, useRef, useState } from 'react';
import menu from '../../assets/images/menu_30.png';
import { useAppDispatch } from '../../app/hooks'

import userAction from "../../actions/user/actions";
import { Link } from 'react-router-dom';

const DropdownMenu: React.FC = () => {
  const dropMenuRef = useRef<HTMLImageElement>(null);
  const dispatch = useAppDispatch();
  const handleLogout = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    dispatch(userAction.logout());
  }
  const [menuOpen, setMenuOpen] = useState(false)
  const handleOpen = () => {
    setMenuOpen(!menuOpen);
  }
  useEffect(() => {

    const handleOutsideClose = (e: any) => {
      e.preventDefault();
      // useRef current에 담긴 엘리먼트 바깥을 클릭 시 드롭메뉴 닫힘
      if (menuOpen && (!dropMenuRef.current?.contains(e.target))) setMenuOpen(false);
    }

    document.addEventListener('click', handleOutsideClose);
    return () => document.removeEventListener('click', handleOutsideClose);
  }, [menuOpen])
  return (
    <>
      <img src={menu} alt='menu' ref={dropMenuRef} onClick={handleOpen} />
      <div className='drop-menu' style={{ display: menuOpen ? 'block' : 'none' }}>
        <ul>
          <li><Link to="/movies">인기영화</Link></li>
          <li><Link to="/fortune">운세마법사</Link></li>
          <li onClick={handleLogout}>로그아웃</li>
        </ul>
      </div>
    </>
  )
}
export default DropdownMenu;