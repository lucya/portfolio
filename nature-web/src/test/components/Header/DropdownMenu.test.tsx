import { render, screen } from '@testing-library/react';
import "@testing-library/jest-dom"
import { Provider } from 'react-redux';
import { HashRouter } from 'react-router-dom';
import store from '../../../app/store';
import DropdownMenu from "../../../components/Header/DropdownMenu";


describe('DropdownMenu', () => {
  test('renders DropdownMenu component', () => {
    render(
      <Provider store={store}>
        <HashRouter>
          <DropdownMenu />
        </HashRouter>
      </Provider>
    );
    const menuElement = screen.getByText(/인기영화/i);
    expect(menuElement).toBeInTheDocument();
  });

  test('renders DropdownMenu link', () => {
    render(
      <Provider store={store}>
        <HashRouter>
          <DropdownMenu />
        </HashRouter>
      </Provider>
    );

    expect(screen.getAllByRole('img')[0]).toHaveAttribute('alt', 'menu')

  });
})