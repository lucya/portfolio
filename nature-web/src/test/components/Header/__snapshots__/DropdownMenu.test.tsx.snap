// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DropdownMenu link 1`] = `
<div>
  <img
    alt="menu"
    src="menu_30.png"
  />
  <div
    class="drop-menu"
    style="display: none;"
  >
    <ul>
      <li>
        <a
          href="#/movies"
        >
          인기영화
        </a>
      </li>
      <li>
        <a
          href="#/fortune"
        >
          운세마법사
        </a>
      </li>
      <li>
        로그아웃
      </li>
    </ul>
  </div>
</div>
`;
